# Encoding Operations
This section contains;
* Rare Encoder
* Label Encoder
* One Hot Encoder

## Data Preparation

In [1]:
from pyspark.sql import SparkSession, functions as F
import findspark

findspark.init("C:\Program Files\Spark\spark-3.3.1-bin-hadoop3")

In [2]:
import warnings

warnings.filterwarnings('ignore')

In [3]:
spark = SparkSession.builder \
    .appName("House Price Model") \
    .master("local[2]") \
    .config("spark.sql.shuffle.partitions", "2") \
    .getOrCreate()

In [4]:
path = "file:///Users/talha/OneDrive/Masaüstü/Talha Nebi Kumru/Data Science/Miuul/Machine Learning/House_Price_Prediction/datasets/outliers/non_outliers.parquet"
df = spark.read \
    .format("parquet") \
    .option("header", True) \
    .option("inferSchema", True) \
    .load(path)

In [5]:
df.limit(5).toPandas()

MSSubClass MSZoning  LotFrontage  ...  SaleType SaleCondition SalePrice
0        60.0       RL         65.0  ...        WD        Normal  208500.0
1        20.0       RL         80.0  ...        WD        Normal  181500.0
2        60.0       RL         68.0  ...        WD        Normal  223500.0
3        70.0       RL         60.0  ...        WD       Abnorml  140000.0
4        60.0       RL         84.0  ...        WD        Normal  250000.0

[5 rows x 75 columns]

In [6]:
from _eda_utils import seperate_columns

In [7]:
numeric_cols, categoric_cols = seperate_columns(spark, df.dtypes)

2023-04-10 03:05:03,478 - logging_utils - DEBUG - seperate_columns function executing...
2023-04-10 03:05:03,481 - logging_utils - INFO - 
Numerical Columns: ['BsmtHalfBath', 'TotRmsAbvGrd', 'YrSold', '3SsnPorch', 'GarageArea', 'LotFrontage', 'YearBuilt', 'BsmtFullBath', 'PoolArea', 'BsmtFinSF2', 'BsmtUnfSF', '1stFlrSF', 'FullBath', 'OverallCond', 'YearRemodAdd', 'WoodDeckSF', 'MoSold', '2ndFlrSF', 'MiscVal', 'OverallQual', 'TotalBsmtSF', 'BsmtFinSF1', 'OpenPorchSF', 'EnclosedPorch', 'LowQualFinSF', 'GarageYrBlt', 'LotArea', 'GrLivArea', 'ScreenPorch', 'MSSubClass', 'BedroomAbvGr', 'HalfBath', 'MasVnrArea', 'GarageCars', 'Fireplaces', 'SalePrice', 'KitchenAbvGr']
2023-04-10 03:05:03,482 - logging_utils - INFO - 
Categorical Columns: ['MSZoning', 'Street', 'LotShape', 'LandContour', 'Utilities', 'LotConfig', 'LandSlope', 'Neighborhood', 'Condition1', 'Condition2', 'BldgType', 'HouseStyle', 'RoofStyle', 'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType', 'ExterQual', 'ExterCond', 'Fo

In [8]:
from _sklearn_linear_regression import LinearRegressionSklearn

In [11]:
def get_score(df, target_col, spark=1):
    if spark: df = df.toPandas()
    reg = LinearRegressionSklearn(df, target_col='SalePrice')
    
    reg.initialize()
    
    reg.model()
    
    return reg.get_score()

In [12]:
get_score(df, 'SalePrice')

0.9096735738461703

## Rare Encoder

### Rare Analyser

In [13]:
from _encoders_utils import rare_analyser

In [14]:
rare_analyser(df, target='SalePrice', categoric_cols=categoric_cols)

2023-04-10 03:05:42,714 - logging_utils - DEBUG - rare_analyser function is executing...
+--------+-----+-----+-----------+
|MSZoning|Count|Ratio|Target Mean|
+--------+-----+-----+-----------+
|      RL| 1141| 0.79|  190518.73|
|      RM|  217| 0.15|  126645.48|
|      FV|   62| 0.04|  214888.13|
| C (all)|   10| 0.01|    74528.0|
|      RH|   16| 0.01|  131558.38|
+--------+-----+-----+-----------+

+------+-----+-----+-----------+
|Street|Count|Ratio|Target Mean|
+------+-----+-----+-----------+
|  Pave| 1441|  1.0|  180766.85|
|  Grvl|    5|  0.0|   110438.6|
+------+-----+-----+-----------+

+--------+-----+-----+-----------+
|LotShape|Count|Ratio|Target Mean|
+--------+-----+-----+-----------+
|     IR2|   41| 0.03|  239833.37|
|     Reg|  915| 0.63|  164256.78|
|     IR1|  481| 0.33|  206078.44|
|     IR3|    9| 0.01|  198373.89|
+--------+-----+-----+-----------+

+-----------+-----+-----+-----------+
|LandContour|Count|Ratio|Target Mean|
+-----------+-----+-----+-----------+
|

### Rare Encoding

In [16]:
from _encoders_utils import rare_encoder

In [17]:
temp_df = df

In [18]:
temp_df = rare_encoder(temp_df, columns=categoric_cols, rare_percent=0.01)

2023-04-10 03:06:28,039 - logging_utils - DEBUG - rare_analyser function is executing...
2023-04-10 03:06:28,107 - logging_utils - INFO - MSZoning column is encoding...
2023-04-10 03:06:29,102 - logging_utils - INFO - Street column is encoding...
2023-04-10 03:06:29,916 - logging_utils - INFO - LotShape column is encoding...
2023-04-10 03:06:30,698 - logging_utils - INFO - LandContour column is encoding...
2023-04-10 03:06:31,545 - logging_utils - INFO - Utilities column is encoding...
2023-04-10 03:06:32,342 - logging_utils - INFO - LotConfig column is encoding...
2023-04-10 03:06:33,106 - logging_utils - INFO - LandSlope column is encoding...
2023-04-10 03:06:33,864 - logging_utils - INFO - Neighborhood column is encoding...
2023-04-10 03:06:34,638 - logging_utils - INFO - Condition1 column is encoding...
2023-04-10 03:06:35,453 - logging_utils - INFO - Condition2 column is encoding...
2023-04-10 03:06:36,285 - logging_utils - INFO - BldgType column is encoding...
2023-04-10 03:06:37

#### Test Rare Encoded

In [19]:
get_score(temp_df, 'SalePrice')

0.9113329100844768

### Save

In [20]:
df = temp_df

In [21]:
df.toPandas().to_parquet('datasets/rare/rare_encoded.parquet', index=False)

### Conclusion

<div class="alert alert-block alert-info"> <b>Final:</b> After rare encoding, accuracy increased <b>0.9096735738461703<b> to <b>0.9113329100844768<b> </div>

## Label Encoding & One Hot Encoding

In [22]:
binary_cols = [col for col in categoric_cols if df.select(col).distinct().count() == 2]
one_hot_cols = [col for col in categoric_cols if col not in binary_cols]

In [23]:
len(binary_cols) + len(one_hot_cols) == len(categoric_cols)

True

### Label Encoding

In [24]:
from _encoders_utils import label_encoder

In [25]:
temp_df = df.toPandas()

In [26]:
temp_df = label_encoder(temp_df, binary_cols)

In [27]:
temp_df[binary_cols].head()

Street  Utilities  Condition2  RoofMatl  Heating  CentralAir
0       0          0           0         0        0           1
1       0          0           0         0        0           1
2       0          0           0         0        0           1
3       0          0           0         0        0           1
4       0          0           0         0        0           1

In [28]:
temp_df.shape

(1446, 75)

In [30]:
df.count()

1446

In [32]:
df = temp_df

### One Hot Encoding

In [33]:
import pandas as pd

In [42]:
temp_df = df.copy()

In [43]:
temp_df = pd.get_dummies(temp_df, columns=one_hot_cols, drop_first=True)

In [44]:
temp_df.head()

MSSubClass  LotFrontage  ...  SaleCondition_Partial  SaleCondition_Rare
0        60.0         65.0  ...                      0                   0
1        20.0         80.0  ...                      0                   0
2        60.0         68.0  ...                      0                   0
3        70.0         60.0  ...                      0                   0
4        60.0         84.0  ...                      0                   0

[5 rows x 171 columns]

In [45]:
df.shape

(1446, 75)

In [46]:
temp_df.shape

(1446, 171)

In [47]:
df = temp_df

## Final Test

In [48]:
get_score(df, 'SalePrice', 0)

0.9113344671996515

In [ ]:
df.to_parquet("datasets/encode/encoded.parquet", index=False)

## Conclusion

<div class="alert alert-block alert-info"> <b>Final:</b> Final encoding, accuracy increased <b>0.9096735738461703<b> to <b>0.9113329100844768<b> </div>

In [49]:
spark.stop()